Source/JavaScriptCore/builtins/ModuleLoader.js

```diff
@@ -23,586 +23,642 @@
 }
 
 @visibility=PrivateRecursive
-async function loadAndEvaluateModule(moduleName, parameters, fetcher)
+function loadModule(key, sourceCode, fetcher)
+{
+    "use strict";
+
+    var entry = this.ensureModuleMapEntry(key, @ModuleTypeJavaScript);
+    entry.fetchAndParsePromise ??= this.parseModule(key, sourceCode).then(moduleRecord => this.hostLoadImportedModule(entry, moduleRecord), error => {
+        entry.errorToRethrow = error;
+        return entry;
+    });
+
+    return @ensureImportMapsLoaded(() => {
+        return entry.fetchAndParsePromise
+            .then(entry => {
+                if (entry.errorToRethrow)
+                    throw entry.errorToRethrow;
+                return entry;
+            })
+            .then(() => this.fetchDescendantsOfAndLink(entry, fetcher));
+    });
+}
+
+@visibility=PrivateRecursive
+function loadModuleAndEvaluate(key, sourceCode, fetcher)
 {
     "use strict";
 
-    var importMap = @importMapStatus();
-    if (importMap)
-        await importMap;
-    var key = this.resolve(moduleName, @undefined, fetcher);
-    key = await this.loadModule(key, parameters, fetcher);
-    return await this.linkAndEvaluateModule(key, fetcher);
+    return this.loadModule(key, sourceCode, fetcher).then(() => {
+        var entry = this.moduleMap.@get(key);
+        @assert(!!entry);
+        return this.moduleEvaluation(entry, fetcher);
+    });
 }
 
 @visibility=PrivateRecursive
-async function requestImportModule(moduleName, referrer, parameters, fetcher)
+function requestImportModule(specifier, referrer, parameters, fetcher)
 {
     "use strict";
 
-    var importMap = @importMapStatus();
-    if (importMap)
-        await importMap;
-    var key = this.resolve(moduleName, referrer, fetcher);
-    var entry = await this.requestSatisfy(this.ensureRegistered(key), parameters, fetcher, new @Set);
-    await this.linkAndEvaluateModule(entry.key, fetcher);
-    return this.getModuleNamespaceObject(entry.module);
+    return @ensureImportMapsLoaded(() => {
+        var key = this.resolve(specifier, referrer, fetcher);
+
+        return this.fetchModuleAndEvaluate(key, parameters, fetcher).then(() => {
+            var entry = this.getModuleMapEntry(key, this.getModuleType(parameters));
+            @assert(!!entry);
+            return this.getModuleNamespaceObject(entry.moduleRecord);
+        });
+    });
 }
 
 @visibility=PrivateRecursive
 function dependencyKeysIfEvaluated(key)
 {
     "use strict";
 
-    var entry = this.registry.@get(key);
-    if (!entry || !entry.evaluated)
-        return null;
+    var entry = this.moduleMap.@get(key);
+    if (!entry || entry.status !== @ModuleStatusEvaluated)
+        return;
 
-    var dependencies = entry.dependencies;
-    var length = dependencies.length;
-    var result = new @Array(length);
-    for (var i = 0; i < length; ++i)
-        result[i] = dependencies[i].key;
+    var result = [];
+
+    entry.loadedModules.@forEach(loadedEntry => {
```

Comment:
Do not use `@forEach`. Use normal iteration.

Reply:
`entry.loadedModules` is a `Map`, we cannot iterate it otherwise (`for/of` would be *actually* observable).